### ISSUE TITLE
caddy not logging basic authentication failures

### ISSUE THREAD

**Issue by sumnerboy12** [caddyserver/caddy#3363]
relates to #2402 which was implemented Feb 2019 pre-v2.

when using `basicauth` to authenticate it would be nice to use `fail2ban` to detect multiple failed login attempts and ban ip addresses. this is difficult with the current logging since every basic authentication attempt starts with a `401` and thus banning on this results in many false positives.

see #2402 for further details and reasoning.

**Comment by mholt** [comment]
Can you clarify more? 

> every basic authentication attempt starts with a 401

This is the correct behavior, unless correct credentials are provided.

**Comment by sumnerboy12** [comment]
yes - agree it is the correct behaviour, and then if an actual authentication attempt fails we also get a 401 (again correct). 

however trying to configure `fail2ban` to ban on just the 401s results in false positives since every time an authentication attempt is started a 401 is logged and this eventually triggers the ban. 

so ideally we have some specific _auth failed_ logging which `fail2ban` can monitor instead of the 401s.

hope that is a little clearer...

**Comment by mholt** [comment]
> if an actual authentication attempt fails

What do you mean by this, as opposed to something else?

Try this: please give me a config, then give me a sequence of `curl` commands, and highlight the behavior that you see, vs. the behavior that you expect to see.

**Comment by sumnerboy12** [comment]
ok - so I have a site authenticated with `basicauth` and I want to use `fail2ban` to ban an ips if there are 3 failed authentication attempts in the last 10 mins (for example).

scenario 1 (valid user);
1. attempts to connect - 401 on initial connect, then auth prompt shown, user enters creds and is authenticated
2. logs out and attempts to reconnect - another 401 on initial connect, then auth prompt, authenticated etc
3. logs out and attempts to reconnect - another 401 <-- at this point `fail2ban` will ban the ip as there have been 3 x 401s in the last few mins (not what we want)

scenario 2 (bad actor);
1. attempts to connect - 401 on initial connect, then auth prompt shown, invalid creds entered, another 401 <-- ideally this would generate some specific _auth failed_ logging which we can then use in our `fail2ban` filter so we only ban on actual authentication failures, and not the initial 401s when starting basic auth

in summary it is providing a mechanism via the caddy logs to identify failed authentication attempts, as apart from the initial 401s logged when basic auth is initiated.

**Comment by mholt** [comment]
Hmm, I think I see what you mean, but I still fail to see why this is a problem.

To clarify: HTTP and basicauth are stateless -- there is no notion of "start" auth and "finish" auth. There is no notion of "second auth attempt" (unless we manage the state server-side, but that requires cookies and all sorts of fun problems).

If users are often "logging out" and "logging in", can't you just raise your fail2ban limit?

**Comment by sumnerboy12** [comment]
I see what you mean - how was this solved in #2404 then? I assumed the same thing could be done in the v2 code (sorry I am not a Go programmer... yet)

**Comment by mholt** [comment]
#2404 has nothing to do with it -- did you mean #2434?

I'm not sure, to be honest.

In your scenario 2, how would you distinguish between a bad actor and a good actor who just entered the wrong password?

**Comment by sumnerboy12** [comment]
gah, sorry #2402...

**Comment by sumnerboy12** [comment]
> #2404 has nothing to do with it -- did you mean #2434?
> 
> I'm not sure, to be honest.
> 
> In your scenario 2, how would you distinguish between a bad actor and a good actor who just entered the wrong password?

this is where the `fail2ban` config (`maxretry`) comes in - if you fail to login say 3 times in the last 10 mins then you will be banned. so a password mis-type won't get you banned...

**Comment by mholt** [comment]
What if they're like me and type their password wrong twice? :P

I guess I'm just trying to see the need to motivate a change in Caddy. AFAICT everything is working as expected, it's just that your fail2ban is too strict probably.

**Comment by sumnerboy12** [comment]
yes, I can see where you are coming from - but as an example I have a service I expose to the internet and have an app that authenticates each time I open it - I sometimes open/close the app multiple times in a row (for whatever reason) - so I would need a very high retry count in `fail2ban` to ensure my mobile device isn't locked out.

I guess I am coming from the POV that I would like to use `basicauth` on a few of my external facing services, and would like a nice easy way to monitor failed login attempts and ban ips as quickly as possible. monitoring 401s alone isn't very reliable/accurate.

but I understand your desire to keep caddy uncluttered. i was hoping this would be a simple change, similar to #2402 but it appears that isn't the case.

**Comment by mholt** [comment]
Hm, so is the problem that you just don't want requests without credentials to be logged? Is that to assume that the first legitimate request won't have credentials and then the following ones will? If so, how would you block bad actors who are just spamming your endpoint with missing credentials? i.e. why should we treat the missing credentials as a special case? What benefit is there to that as opposed to just raising the threshold by 1?

(Sorry for all the questions. I want to make sure I understand.)

**Comment by sumnerboy12** [comment]
raising the threshold doesn't help, because you still get 401s when successfully authenticating (on the initial connect). so if you authenticate multiple times you will be banned (assuming authenticate count > threshold).

looking at #2434 I think the only change was to add some logging to *all* 401s raised by `basicauth`, which included the `username`. if that was added to v2 then I guess we could just configured `fail2ban` to ignore 401s with no (or `null`) username and only ban if a username is being used.

**Comment by sumnerboy12** [comment]
> (Sorry for all the questions. I want to make sure I understand.)

Please don't apologise! I very much appreciate your help with this - I have just moved across to Caddy and am loving it - fantastic piece of software!

**Comment by mholt** [comment]
@sumnerboy12 How about if we emit debug logs for every basicauth attempt, that has the username in the log?

**Comment by sumnerboy12** [comment]
that would work I think - that sounds pretty similar to what was done for #2434 - so would at least be inline with that.

**Comment by mholt** [comment]
@sumnerboy12 Can you please try the commit on the `auth-log` branch, 1103a19  -- there will be build artifacts available soon you can download and use.

Make sure to enable debug logging and you'll see it.

**Comment by dmknob** [comment]
> yes - agree it is the correct behaviour, and then if an actual authentication attempt fails we also get a 401 (again correct).

@mholt 
When an actual authentication attempt fails, 403 Forbidden isn't more suitable than 401 Unauthorized?
"The client does not have access rights to the content; that is, it is unauthorized, so the server is refusing to give the requested resource. Unlike 401, the client's identity is known to the server." https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/403

**Comment by francislavoie** [comment]
Technically you don't know the identity, because they haven't successfully authenticated. You have a username, but that's not enough to say that you know who it is. `401` is definitely the correct status code to return on a failed authentication.

**Comment by mholt** [comment]
@dmknob No, 403 is an authorization error, not an authentication error (the standardized text of 401 is unfortunate). Re-authenticating (i.e. with fixed credentials) may very well succeed. It's not that the server knows the user is not _authorized_ to view the resource.

**Comment by a10kiloham** [comment]
Is there any good guide yet to implement fail2ban with caddy 2? This seem like a partial fix up thread but not complete...

**Comment by mholt** [comment]
This feature has been abandoned by @sumnerboy12.

**Comment by sumnerboy12** [comment]
I still think this would be a worthwhile change for caddy - but will leave it to you guys to decide.

**Comment by mholt** [comment]
@sumnerboy12 You had the actual use case for it. We're not going to implement a new feature _just because_. Since it is no longer required, it will not be implemented. My time must be spent on things that will make a difference to the project and its users, not "nice to haves" (especially when they take so much time like this feature).

**Comment by sumnerboy12** [comment]
Fair enough - didn't mean to offend or waste anyone's time.

**Comment by mholt** [comment]
No offense taken, but in the future please consider keeping developers updated when you have big changes like that.
